### PR TITLE
Add optional secondary serial device to ESP32 receivers

### PR DESCRIPTION
**Pull request by mha1**
This PR intends to enable more possible use cases mostly for but not limited to fixed wing and helicopter pilots.

As an example consider the classical use cases:
- a fixed wing setup with external CRSF telemetry and an external gyro connected to a receiver
- a helicopter setup with a flybarless unit and external telemetry sensors (e.g. a telemetry capable ESC)

Currently ELRS receivers only allow using one serial interface (I call it the primary one) for the implemented set of protocols (CRSF, CRSF INV, SBUS, SBUS INV, DJI RS PRO HoTT Telemetry) making the above use cases impossible. This PR adds a secondary serial device on ESP32 receivers which enables running an additional protocol. This opens up the possibility to run two different protocols concurrently. In the above uses cases a setup could look like:
- primary serial CRSF, secondary serial SBUS
- primary serial HoTT Telemetry, secondary serial SUMD

Here's an example using a Radiomaster ER6 receiver showing protocol HoTT Telemetry with a GPS/Vario connected on the primary serial (RX and TX used) and SBUS on the secondary serial (only TX used) configured to use the signal pin of channel connector 6 which might be used to feed channel data to a flybarless unit or gyro. ER6 channels 1 to 5 still output PWM data.

pr

The secondary serial device can be configured using the WebUI or LUA script

Extensions to the code base:
- added an additional serial device
- added configuration options to WebUI and LUA
- added logic to run any of the known serial protocols on the secondary serial device

### CODE REVIEW

**Comment by pkendall64** [comment]
The second serial and protocol should be selectable from the PWM page not the hardware page.
We may still keep the second serial pins on the hardware page for non-PWM receivers.

The hardware page should only be used to configure the actual hardware layout. Users should not be going to the hardware page to configure normal features.

**Comment by mha1** [comment]
I first started thinking this might be an expert option but of course I see your point. To do a proper user UI I need some discussion on how to present this and also on how to implement this. scan.js is not self explanatory. Is there a UI guy we could try to pull in?

**Comment by pkendall64** [comment]
> Is there a UI guy we could try to pull in?
That would be me as well 😭

**Comment by mha1** [comment]
Example: SuperP 14CH configured to have SBUS on the primary serial and HoTT telemetry on the primary serial defined to be on channel 11 and 12 (JST connector):

image
image

2: ch11/12 on JST
3: ch14

image

**Comment by JyeSmith** [comment]
Protocol2 get my vote.

**Comment by mha1** [comment]
names updated

**Comment by mha1** [comment]
This is what it looks like currently. Let me know if you have other ideas.

image
image
image
image
image

**Comment by 55Laurent** [comment]
For your info some feedback, I made some tests of this feature on RM ER6 and Betafpv super P + a graupner +T70 esc and I also checked the Sbus output with a MC8 checker. 
Everything work fine, telemetry is ok (I got the same sensors data that I can get with a genuin graupner receiver) and the Sbus is also working fine with both receivers. I used the web UI to set serial interfaces (Hott telem and Sbus), I also checked that thoses setting can be done with the lua srcipt. 
Anything is perfect and very handy.  Very nice job

File: src/html/hardware.html [hunk 1]
```diff
@@ -86,6 +86,10 @@ <h1><b>ExpressLRS</b></h1>
 					<tr><td width="30"></td><td>RX pin<img class="icon-input"/></td><td><input size='3' id='serial_rx' name='serial_rx' type='text'/></td><td>Pin used to receive CRSF signal from the handset</td></tr>
 					<tr><td></td><td>TX pin<img class="icon-output"/></td><td><input size='3' id='serial_tx' name='serial_tx' type='text'/></td><td>Pin used to transmit CRSF telemetry to the handset (may be the same as the RX PIN)</td></tr>
 
+					<tr><td colspan='2'><b>Serial2 Pins</td></tr>
+					<tr><td width="30"></td><td>RX pin<img class="icon-input"/></td><td><input size='3' id='serial1_rx' name='serial1_rx' type='text'/></td><td>Serial2 RX - ESP32 targets only</td></tr>
+					<tr><td></td><td>TX pin<img class="icon-output"/></td><td><input size='3' id='serial1_tx' name='serial1_tx' type='text'/></td><td>Serial2 TX - ESP32 targets only</td></tr>
```
Comment (pkendall64) [inline src/html/hardware.html:91]: You will need to add `serial1_rx` & `serial1_tx` to the target validator in the targets repo.
See https://github.com/ExpressLRS/targets/blob/master/.github/hardware.py

File: src/html/scan.js [hunk 2]
```diff
@@ -375,15 +399,37 @@ function updateConfig(data, options) {
...
+      return;
+    }
+  }
+
   updatePwmSettings(data.pwm);
   _('serial-protocol').value = data['serial-protocol'];
   _('serial-protocol').onchange();
+  _('serial1-protocol').value = data['serial1-protocol'];
+  _('serial1-protocol').onchange();
   _('is-airport').onchange = _('serial-protocol').onchange;
```
Comment (pkendall64) [inline src/html/scan.js:418]: `is_airport` onchange should be function that now calls both serial and serial1 onchange functions.

**Review by pkendall64 — COMMENTED** [review]
You will also need to provide a config upgrader

File: src/html/hardware.html (hunk 1 above)
Comment (mha1) [reply src/html/hardware.html:91]: please review https://github.com/ExpressLRS/targets/pull/60

File: src/html/scan.js (hunk 2 above)
Comment (mha1) [reply src/html/scan.js:418]: is_airport: done
upgrader: done, but not sure about this. initializing serial1 protocol should be enough, right?

**Comment by pkendall64** [comment]
@mha1 One oddity I noticed is that if I set Serial2RX on pin13 and then Serial2TX on pin 14, pin 13 goes back to 50Hz in the web UI. Probably not a big deal.

**Review by pkendall64 — APPROVED** [review]

**Comment by mha1** [comment]
> @mha1 One oddity I noticed is that if I set Serial2RX on pin13 and then Serial2TX on pin 14, pin 13 goes back to 50Hz in the web UI. Probably not a big deal.

ch13/14 on the SuperP?

**Comment by pkendall64** [comment]
> > @mha1 One oddity I noticed is that if I set Serial2RX on pin13 and then Serial2TX on pin 14, pin 13 goes back to 50Hz in the web UI. Probably not a big deal.
> 
> ch13/14 on the SuperP?

Yes, SuperP sorry.

**Comment by mha1** [comment]
> > > @mha1 One oddity I noticed is that if I set Serial2RX on pin13 and then Serial2TX on pin 14, pin 13 goes back to 50Hz in the web UI. Probably not a big deal.
> > 
> > 
> > ch13/14 on the SuperP?
> 
> Yes, SuperP sorry.

Thanks for pointing out this problem. This needed a little features logic refinement. If PWM pins are already featured as Serial RX and TX due to a fixed rx/tx assignment in hardware.html you shouldn't be able to assign Serial2 RX or TX to those pins. Means if PWM pins are already featured as Serial1 they can't be featured as Serial 2. Updated logic accordingly which will prevent selecting Serial 2 RX and/or TX on a Serial 1 featured pins.

No more Serial2 feature badge for receivers having dedicated Serial 1 PWM pins:

image

**Comment by JyeSmith** [comment]
Looks good.

I tested by defining the RP4 second uart with the below hardware setup.  I was then able to set a combination of CRSF, SBUS, SUMD protocols on the 2 uarts via the Rx Lua.  Switching between them I was able to connect to BF without an issue through any receiver uart and protocol.

FYI this opens the door for VTs protocols... Get to it! 

image

image

**Review by JyeSmith — APPROVED** [review]

**Pull request merged**